src/shogun/base/SGObject.h

```diff
@@ -373,6 +373,7 @@ class CSGObject
 			}
 			ref_value(value);
 			update_parameter(_tag, make_any(value));
+			observe(ObservedValue::make_observation<T>(1, _tag.name(), get_parameter(_tag)));
```

Comment:
This line here has a little problem. Since `ObservedValue` is a forward declaration, I cannot access its methods/constructor. However, I cannot `#include ObservedValue.h` here since it causes a bit of a mess with SGObject itself (it is a cyclic dependency problem). Moreover, since this method is templated I can't place it into a `.cpp` file. Therefore, this line as it is does not compile. :(

Any suggestions about how to fix it? @karlnapf @vigsterkr @lisitsyn

Reply:
Move into a namespace method that accepts the SGObject as input

Reply:
ehm I actually cannot see exactly the solution you're proposing. Could you elaborate a bit more?

Reply:
You would not call the ctor/method here (so you need the definitions) but rather a method that just returns what you want (so forward declaration is ok) and define this method in another file to avoid the cyclic dependency.

Reply:
I've tried a bit yesterday evening but I think it is not the proper solution to this problem. Since `make_observation` is templated, even if I create another file for it, it is not possible to split the declaration and definition. Moreover, even if I were to create a method like this (which indeed compiles since we use only pointers):
```c++
ObservedValue * make_observation(...)
{
    return new ObservedValueTemplated<T>(...);
}
```
I would still need to include `<shogun/lib/parameter_observer/ParameterObserverTemplated.h>` in each cpp file which uses SGObject (basically 99% of Shogun's files), since otherwise the compiler will not be able to resolve the forward declaration of `ObservedValueTemplated`.

Reply:
I think you might be able to define a templated function in SGObject.h, and then put the implementation into the SGObject.cpp file and use `SG_FORCE_INLINED`. Then you can include the observed value header in SGObject.cpp while forward declaring it in SGObject.h

Reply:
Just tried this:

SGObject.cpp
```
namespace shogun
{
template <class T>
SG_FORCED_INLINE void test()
{
	T* a=nullptr;
}
}
```

SGObjecth
```
template <class T>
void test();
```

and then call the function e.g. inside `CSGObject::get<T>` which compiles and links fine.

Reply:
I've just tried it in every possible manner to make it works, but, in the end, we will have always linker errors (especially when one tries to use the library). That's happens even if the method is represented as inline.